### PR TITLE
Reorganize RequestMapper and friends

### PR DESCRIPTION
**Pull request by slinkydeveloper**
The goal of this PR is to properly split `RequestMapper` into different classes, one for each concern. The final result of this PR looks like this:

* The part of `RequestMapper` doing the reconciliation of the internal maps of producers is moved inside the class called `IngressProducerReconciliableStore`.
* The logic of the request handling is not organized anymore in decorators and then `RequestMapper` as the final step, but it's "orchestrated" by the `ReceiverVerticle.handle` method as follows:
  1. A bunch of "pre handlers" are executed, these includes the method not allowed check, the probe handler and potentially other checks/headers put in future, like the webhoook stuff. These just implement `Handler<HttpServerRequest>` and do not wrap the next handler anymore.
  2. The `ReceiverVerticle` owns the `IngressProducerReconciliableStore`, which it queries to find out the `IngressProducer` associated to that request
  3. If the `IngressProducer` resolved is not null, the `IngressRequestHandler` is invoked, which will produce the record and end the request

### CODE REVIEW

**Comment by slinkydeveloper** [comment]
cc @pierDipi how does it looks like? I still need to reorg the tests, but is it a bit more organized now?

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/handler/MethodNotAllowedHandler.java
Comment (matzew) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/handler/MethodNotAllowedHandler.java:52]: I think we should allow `OPTIONS` too, see here: https://github.com/knative/eventing/pull/5542

and perhaps also set the `Allow` header (see PR above too for this)

Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/handler/MethodNotAllowedHandler.java:52]: Another PR and another handler singleton :smile:

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java
Comment (matzew) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java:39]: wanna add some javadoc here?

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java
Comment (matzew) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java:105]: when would we hit this?

**Comment by matzew** [comment]
General I like the notion of `Ingress...` on the class names. We should do for "egress" too (e.g. `Egress...`).

Just wondering should we also move that to dedicated packages?

**Comment by slinkydeveloper** [comment]
> Just wondering should we also move that to dedicated packages?

About package reorg, I think what we have in this PR is already good enough... Do you have something in particular?

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java
Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java:105]: if you're doing a request to a broker path that does not exists

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java
Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java:50]: Added some details from the PR description

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/handler/MethodNotAllowedHandler.java
Comment (matzew) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/handler/MethodNotAllowedHandler.java:52]: OK. makes sense. I'd like to send that handler for this - OK ?

File: data-plane/receiver/src/test/java/dev/knative/eventing/kafka/broker/receiver/handler/IngressRequestHandlerImplTest.java
Comment (pierDipi) [inline data-plane/receiver/src/test/java/dev/knative/eventing/kafka/broker/receiver/handler/IngressRequestHandlerImplTest.java:1]: Copyright

File: data-plane/receiver/src/test/java/dev/knative/eventing/kafka/broker/receiver/handler/PreHandlerTest.java
Comment (pierDipi) [inline data-plane/receiver/src/test/java/dev/knative/eventing/kafka/broker/receiver/handler/PreHandlerTest.java:1]: Copyright

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java
Comment (pierDipi) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java:109]: What if a preHandler handles a request asynchronously, would this return false?
It seems we can a build a more robust design here :thinking:

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/IngressRequestHandler.java
Comment (pierDipi) [inline data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/IngressRequestHandler.java:25]: Nice!

**Review by pierDipi — COMMENTED** [review]
I like this! I left a few comments.

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java
Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java:109]: Do we have an actual use case for that? I feel like I'm recreating `vertx-web`, and I don't wanna go down that route :smile: How about we introduce that change only if a use case comes out?

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/handler/MethodNotAllowedHandler.java
Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/handler/MethodNotAllowedHandler.java:52]: ok, after this pr is merged feel free to do so.

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java
Comment (pierDipi) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java:109]: A class shouldn't change because another one changes. 
Why don't we organize these handlers as a chain https://en.wikipedia.org/wiki/Chain-of-responsibility_pattern?

Comment (slinkydeveloper) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java:109]: > A class shouldn't change because another one changes.

Of course, what I meant is that because we have no use case for async pre handlers, nor I can think of any future use cases in mind for that, there is no need for `ReceiverVerticle` to support async handlers, because it makes the code more complicated for no reason. 

What I'm implementing here is already a chain of responsability, where handlers are invoked one after the other where each one does its job, It's just not meant to include async handlers. In order to implement it with async handlers, we should have `Function<HttpServerRequest, Future<Void>>`.

The reason why the `IngressRequestHandler` is "special" is because I would like to have the implementation be unaware of the store, the reconciliation and all the complexity brought in with it. I tried to find a nice way to fix that, but then the injection of the `vertx` object would become too much complicated, so I'm out of solutions atm :disappointed: 

The solution in this PR is not be best solution of course, but I guess it's a beginning...

**Review by matzew — APPROVED** [review]
/lgtm
/hold


for other to jump in

**Comment by pierDipi** [comment]
There are failures https://github.com/knative-sandbox/eventing-kafka-broker/pull/1041/checks?check_run_id=2931263174
```
Error:  Errors: 
Error:    RequestMapperTest.lambda$testStore$18:298->lambda$testStore$15:298 » NullPointer
Error:    RequestMapperTest.lambda$testStore$18:298->lambda$testStore$15:298 » NullPointer
Error:    RequestMapperTest.lambda$testStore$18:298->lambda$testStore$15:298 » NullPointer
Error:    RequestMapperTest.lambda$testStore$18:298->lambda$testStore$15:298 » NullPointer
Error:    RequestMapperTest.lambda$testStore$18:298->lambda$testStore$15:298 » NullPointer
```

File: data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java
Comment (pierDipi) [reply data-plane/receiver/src/main/java/dev/knative/eventing/kafka/broker/receiver/impl/ReceiverVerticle.java:109]: I see, let's keep this for now

**Pull request merged**
